services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -114,13 +116,16 @@ private void doDownloadDirectory(CompletableFuture<CompletedDirectoryDownload> r
         Queue<FailedFileDownload> failedFileDownloads = new ConcurrentLinkedQueue<>();
 
         CompletableFuture<Void> allOfFutures = new CompletableFuture<>();
-        AsyncBufferingSubscriber<S3Object> asyncBufferingSubscriber =
-            new AsyncBufferingSubscriber<>(s3Object -> downloadSingleFile(downloadDirectoryRequest,
-                                                                          failedFileDownloads,
-                                                                          s3Object),
+        AsyncBufferingSubscriber<Pair<S3Object, Path>> asyncBufferingSubscriber =
+            new AsyncBufferingSubscriber<>(objectAndPath -> downloadSingleFile(downloadDirectoryRequest,
+                                                                               failedFileDownloads,
+                                                                               objectAndPath),
```

Comment:
I like the pairing of object and path (don't like the name but can't think of a better one; it IS descriptive) which allows for a more efficient implementation of the the single file download; despite the new functionality the class only grew by 3 lines.